sdks/java/core/src/test/java/org/apache/beam/sdk/options/ValueProviderTest.java

```diff
@@ -0,0 +1,214 @@
+    TestOptions runtime = mapper.readValue(
+      "{ \"options\": { \"foo\": \"quux\" }}", PipelineOptions.class)
+      .as(TestOptions.class);
+
+    TestOptions options = PipelineOptionsFactory.as(TestOptions.class);
+    runtime.setOptionsId(options.getOptionsId());
+    RuntimeValueProvider.setRuntimeOptions(runtime);
+
+    ValueProvider<String> provider = options.getFoo();
+    assertFalse(provider.shouldValidate());
+    assertEquals("quux", provider.get());
+  }
+
+  @Test
+  public void testDefaultRuntimeProviderWithOverride() throws Exception {
+    ObjectMapper mapper = new ObjectMapper();
+    TestOptions runtime = mapper.readValue(
+      "{ \"options\": { \"bar\": \"quux\" }}", PipelineOptions.class)
+      .as(TestOptions.class);
+
+    TestOptions options = PipelineOptionsFactory.as(TestOptions.class);
+    runtime.setOptionsId(options.getOptionsId());
+    RuntimeValueProvider.setRuntimeOptions(runtime);
+
+    ValueProvider<String> provider = options.getBar();
+    assertTrue(provider.shouldValidate());
+    assertEquals("quux", provider.get());
+  }
+
+  /** A test interface. */
+  public static interface BadOptionsRuntime extends PipelineOptions {
+    RuntimeValueProvider<String> getBar();
+    void setBar(RuntimeValueProvider<String> bar);
+  }
+
+  @Test
+  public void testOptionReturnTypeRuntime() {
+    BadOptionsRuntime options = PipelineOptionsFactory.as(BadOptionsRuntime.class);
+    expectedException.expect(RuntimeException.class);
+    expectedException.expectMessage(
+      "Method getBar should not have return type "
+      + "RuntimeValueProvider, use ValueProvider instead.");
+    RuntimeValueProvider<String> provider = options.getBar();
+  }
+
+  /** A test interface. */
+  public static interface BadOptionsStatic extends PipelineOptions {
+    StaticValueProvider<String> getBar();
+    void setBar(StaticValueProvider<String> bar);
+  }
+
+  @Test
+  public void testOptionReturnTypeStatic() {
+    BadOptionsStatic options = PipelineOptionsFactory.as(BadOptionsStatic.class);
+    expectedException.expect(RuntimeException.class);
+    expectedException.expectMessage(
+      "Method getBar should not have return type "
+      + "StaticValueProvider, use ValueProvider instead.");
+    StaticValueProvider<String> provider = options.getBar();
+  }
+
+  @Test
+  public void testSerializeDeserializeNoArg() throws Exception {
+    TestOptions submitOptions = PipelineOptionsFactory.as(TestOptions.class);
+    assertFalse(submitOptions.getFoo().shouldValidate());
+    ObjectMapper mapper = new ObjectMapper();
+    String serializedOptions = mapper.writeValueAsString(submitOptions);
+
+    // This is the expected behavior of the runner: deserialize and set the
+    // the runtime options.
+    String anchor = "\"appName\":\"ValueProviderTest\"";
+    assertThat(serializedOptions, containsString("\"foo\":null"));
+    String runnerString = serializedOptions.replaceAll(
+      "\"foo\":null", "\"foo\":\"quux\"");
```

Comment:
Concept proved with the string substitution, but let's provide the JSON-fixing utility so that other runners can use it. Then we can use that utility here. I imagine it takes two arguments:

* `String` serialized options
* `Map<String, String>` runtime values to be added

And then it produces a new `String` serialized-options-with-runtime-values-added. (Maybe -- there are other plausibly right interfaces, but this seems closest to what you expect runners to do in practice.